configfile: "config.yaml"

import io
import os
import yaml
import csv
import pandas as pd
import numpy as np
import pathlib
from snakemake.exceptions import print_exception, WorkflowError 

with open("config.yaml", "r") as configfile:
    config = yaml.load(configfile)
    
with open("cluster.yaml", "r") as clusterfile:
    cluster = yaml.load(clusterfile)

INPUTDIR = config["inputdir"]
REFERENCEDIR = config["referencedir"]
SCRATCHDIR = config["scratch"]
OUTDIR = config["outputdir"]
DATADIR = config["datadir"] # for transcriptomes not in MMETSP
SUBDIRECTORY = config["subdirectory"]
SUBDIR_TABLE = pd.read_csv(SUBDIRECTORY, sep = "\t")
DATE = config["date"]

PN_name = config["pn"]
Alex_name = config["alex"]
Parasite_name = config["amoeb"]
Ehux_name = config["ehux"]
Thaps_name = config["thaps"] 
Heterocap_name = config["het"]
DICT_NAMES = {Heterocap_name:"HT",\
              Thaps_name:"TP",\
              PN_name:"PN",\
              Parasite_name:"AB",\
              Alex_name:"AX",\
              Ehux_name:"EH"} 
DICT_CODES = {v: k for k, v in DICT_NAMES.items()} 

GROUP_NAMES = []

# Creates the salmon indices for the groupings of organisms in
# scRNA-seq SeqWell experiment. 
# 
# @param list_species The organisms that we want in the index 
def make_index(list_species, reverse_name=False):
    string_input = ""
    tg_map_names = ""
    
    if reverse_name == False:
        species_sep = list_species.split("_")
        for species in species_sep:
            string_input = string_input + os.path.join(DATADIR, species + "_combined_nt.fasta") + " "
            tg_map_names = tg_map_names + os.path.join(DATADIR, "tgMap_" + species + ".tsv ")
    else:
        string_input = os.path.join(DATADIR, reverse_name + "_combined_nt.fasta")
        tg_map_names = tg_map_names + os.path.join(DATADIR, "tgMap_" + reverse_name + ".tsv")
    
    print(string_input)
    path_fasta = os.path.join(DATADIR, list_species + ".fa")
    path_tgmap = os.path.join(DATADIR, "tgMap_" + list_species + ".tsv")
    
    # concatenate FASTA files
    os.system("cat " + string_input + " > " + path_fasta)
    
    # create salmon index 
    os.system("salmon index -t " + path_fasta + " -i data/indices/" + list_species + "_index -k 31")        
    salmon_index = os.path.join(DATADIR, "indices", list_species + "_index")    

    # concatenate transcript-to-gene maps
    os.system("cat " + tg_map_names + " > " + path_tgmap)
    
def make_tg_reverse(short_name):
    file_loc = os.path.join(DATADIR, "tgMap_" + short_name + ".tsv")
    os.system("touch " + file_loc)

    with open(file_loc, 'wt') as tgMap_file:
        transcript_to_gene = csv.writer(tgMap_file, delimiter='\t')
        command = str("cat " + os.path.join(DATADIR, str(short_name) + "_combined_nt.fasta") + " | grep \">\" | cut -f2 -d \">\" | cut -f1 -d \" \" > transcript_names.txt")
        os.system(command)
        print(command)
        transcripts = open("transcript_names.txt", "r")
        for transcript in transcripts:
            print(transcript)
            genes = [transcript.replace("\n",""), short_name + "-" + "DN" + transcript.split("_")[1].replace("\n", "")]
            transcript_to_gene.writerow(genes)

    tgMap_file.close()

FILE_NAMES = SUBDIR_TABLE.File
INDEX_NAMES = SUBDIR_TABLE.Index
DIRECTORIES = SUBDIR_TABLE.Directory

list_cultures = INDEX_NAMES
print(list_cultures)

# perform preprocessing steps
import preprocessing

include: "assembly-snake"  
include: "index-snake"    
include: "tgmap-snake"    

rule all:
    input:
        # ASSEMBLY OUTPUTS
        assembly = expand(os.path.join(DATADIR, "{sample}_combined_nt.fasta"), sample = DIRECTORIES),
        # INDEX OUTPUTS
        index = expand(os.path.join(DATADIR, "indices", "{sample}_index"), sample = DIRECTORIES),
        # TG MAP OUTPUTS
        tgmap = expand(os.path.join(DATADIR, "tgMap_{sample}.tsv"), sample = DIRECTORIES)


#for list_species in list_cultures:
#    if config["usereverse"] == 1:
#        make_index(list_species,dir_curr) # now we have a Fasta file and tgMap file in the datadir to work off of
#    else:
#        make_index(list_species)
